### DISCUSSION TITLE
lwip (w5500) lib timing "issues"

### DISCUSSION THREAD

**Discussion by Ing-Dom** [earlephilhower/arduino-pico#1774] [General]
Hi,

I moved my project from "Ethernet_Generic" stack to the core's lwip_w5500 implementation.

I recognized some behaviour that I did not expect and I did not experience before with the Ethernet_Generic stack (application more or the less unchnanged).

My application sends (udp.begin, write, end) and receives (udp.read) data.
I monitor my loop timings closely to ensure my response time.

There are quite often some delays > 10ms, but also some delays from about 50-100ms in rare cases.
These delays occour in functions that cannot be the cause, so I assume that maybe interrupts cause this.
Is that possible?

**Comment by earlephilhower** [comment]
I'm traveling so can't find the exact PR, but if you look at the merged PR that added wired Ethernet you'll see the discussion about this.  The core uses native LWIP and a polled architecture.  You can manually adjust the polling frequency by editing a constant, described there.  It should be possible to implement an IRQ based setup with some work.

There is also probably support for multiple packets per notification when using the cooked stack in the W5100 chip.  Not sure if that's possible in raw Ethernet mode which is what running LWIP needs.

**Comment by Ing-Dom** [comment]
I think you mean this one right?
https://github.com/earlephilhower/arduino-pico/pull/1703

To understand it correctly:
The core is polling the ethernet chip on a fixed freqency and lwip handles the packages. when I do udp.readpackages I only read out some buffer that was previously filled?

Is this polling in a timer interrupt or is it in the main loop ?

The problem is, I see quite big delays (often in range of 5-20ms, less often up to 100ms) when using wired ethernet.
This shouldn't be the case, should it?
I suspect lwip but that is just an educated guess...

**Reply by earlephilhower** [reply]
It's based on the asynchronous handler developed by the Raspberry Pi team, same as is used by the Pico W.

Try adding the lwipPollingPeriod(1); call discussed in that PR to your sketch and see if that helps.

For outgoing packets it goes out more or less immediately.  For incoming you may expect at least 1/2 the polling period delay on average before it's detected.

I'm not sure how the raw packet handler runs in the Wxxx chips, but if it only lets you receive one raw Ethernet packet per poll then that may be limiting since the cooked mode that runs IP in the Wxxx chips can return multiple packets at a time.  Or maybe it's possible to get >1 packet in raw mode but we're not trying.  That would be an easy change to make in the polling routine.

**Comment by Ing-Dom** [comment]
ah now..
I think I did not use the correct term.
I experience that the loop is **blocked** for up 100ms. I wrote delay and I think you understood that the packages are received delayed.
That is not the case.

It is about a blocked main loop.
A few ms are acceptable here I think, but not up to 100ms. 

I'll try to track this down further. But maybe you have some idea or hint for me.

**Reply by earlephilhower** [reply]
I'll be traveling for the next two weeks with limited Internet and no computer, so unfortunately I can't be of much help.  Off the top of my head, I can't think of anything that would freeze the main loop for 1/10th of a second.
